### DISCUSSION TITLE
Is there any way to configure SSH keys to Lazygit?

### DISCUSSION THREAD

**Discussion by hasin023** [jesseduffield/lazygit#3165] [Q&A]
Every time I try to pull or push to a remote repo that is using the **SSH** link, It says, _"Permission denied (publickey)"_. How can I resolve this??

**Comment by mark2185** [comment]
How do you push through the `cli`?

**Reply by hasin023** [reply]
In the most standard way that lazy git has to offer. 'a'[To add all un-tracked files] -> 'c' [To write commit mssg and commit] -> 'P' [to push]

**Reply by mark2185** [reply]
I was thinking through using `git` itself. Do you have to put in the passphrase or stuff like that.

**Reply by samuelnihoul** [reply]
yes we do

**Reply by mark2185** [reply]
@samuelnihoul then look into an SSH agent program like [keychain](https://wiki.archlinux.org/title/SSH_keys#Keychain).

Once you add your key into the keychain you'll only have to enter the passphrase the first time it is used.

**Reply by DusanLesan** [reply]
You can use gnome keyring which can be set up to unlock on login to avoid additional pass prompts for ssh https://wiki.archlinux.org/title/GNOME/Keyring

**Reply by MartianInGreen** [reply]
I hate that the answer everyone gives is just _use an ssh agent_ as if there aren't actual good reasons for needing (or wanting to need) to enter the password every time.

**Comment by RollsChris** [comment]
I also have this issue

**Reply by hasin023** [reply]
I couldn't find any fix, so I just stopped using Lazygit :")

**Comment by hasecilu** [comment]
You need to configure SSH on **YOUR SYSTEM** not Lazygit, maybe the agent is not working. Maybe bad configuration.
I wrote [a guide](https://gitlab.com/hasecilu/dotfiles/-/snippets/2542670) some time ago maybe someone find something useful there.

**Reply by RollsChris** [reply]
I have my SSH-Agent running. I can use SSH repo fine when i use git from powershell... but when i load lazygit it breaks.... why does lazygit do anything special.. shouldn't it just be wrapping around the existing git command? its like its using its own local git command

**Reply by hasecilu** [reply]
Mmm, maybe there is a Windows limitation that prevents the same flawless behaviour on Linux.

**Reply by RollsChris** [reply]
or maybe windows is being snubbed by the linux possy ;)  anyway i think this repo should just remove the statement that it supports windows

**Reply by o-token** [reply]
It also happens on my ubuntu machine.

**Reply by mark2185** [reply]
@SuperSocialForce did you follow @hasecilu's guide and which step fails? 

The general solution is to look into an SSH agent program like [keychain](https://wiki.archlinux.org/title/SSH_keys#Keychain).

**Reply by o-token** [reply]
It worked! Thanks

**Comment by ArcSpecter** [comment]
i have the same issue. lazygit is giving me error: Permission denied (publickey, password)
although from same terminal from where I run lazygit if I push things using Git command line, it works just fine (it asks password, i enter it and it pushes changes in)

and I am Debian user.

i guess lazygit is not working properly here. as simple as that
its ok..back to Sublime Merge ;-)

**Reply by mark2185** [reply]
@r2com Look into an SSH agent program like [keychain](https://wiki.archlinux.org/title/SSH_keys#Keychain).

Once you add your key into the keychain you'll only have to enter the passphrase the first time it is used.

**Comment by EricDuminil** [comment]
For what it's worth:
on my Ubuntu laptop, I have SSH private keys 1 and 2. Both can give access to my git server.
Key 1 is loaded in ssh-agent. `git pull` works fine in terminal, and doesn't ask for any password. It apparently uses Key 1.

When starting lazygit, it asks me for password for Key 2, which isn't required to access the git server anymore.

**Comment by toberoni** [comment]
I had this problem because my project's remote repo was configured to use password auth (I guess by VSCode). 

In your project try `git remote -v` and if you see a URL with https this is likely the problem:
```bash
> origin  https://github.com/YOUR_GIHUB_USERNAME/YOUR_REPO_NAME.git (fetch)
```

**Solution:**
If you change your remote URL the CLI & Lazygit should use SSH instead of password auth:

```bash 
git remote set-url origin git@github.com:YOUR_GIHUB_USERNAME/YOUR_REPO_NAME.git
```

Hope that helps!

**Reply by arieldd** [reply]
This solved it for me, thanks.

**Comment by Souvlaki42** [comment]
Did anyone find a Windows specific solution? It worked just fine when I tested on Linux but my main workspace is on Windows because I have issues with WSL2.

**Reply by mark2185** [reply]
The same solution as above should work here as well - [setting up an SSH agent](https://interworks.com/blog/2021/09/15/setting-up-ssh-agent-in-windows-for-passwordless-git-authentication/).

**Comment by azinsharaf** [comment]
i think i have found why it fails on window. not sure what could be a workaround.
When i create a ssh key (and include a passphrase) then lazygit can't pull. I created a ssh key without a passphrase and lazygit works.

**Reply by mark2185** [reply]
The same solution as above should work here as well - [setting up an SSH agent](https://interworks.com/blog/2021/09/15/setting-up-ssh-agent-in-windows-for-passwordless-git-authentication/).

**Comment by tom-gora** [comment]
Had this problem resolved it writing myself a wrapper function around lazygit that replaced my alias:
``` bash
# wrapper for lazygit to init ssh-agent
lg() {
  local IS_AGENT IS_IDENTIFIED FIRST_AVAILABLE
  IS_AGENT="$(pgrep -u "$USER" ssh-agent)"
  local PRIMARY_SSH_KEY="$HOME/.ssh/github_auth_primary"
  local FALLBACK_SSH_KEY="$HOME/.ssh/github_auth_fallback"

  if [ -f "$PRIMARY_SSH_KEY" ]; then
    FIRST_AVAILABLE="$PRIMARY_SSH_KEY"
  else
    if [ -f "$FALLBACK_SSH_KEY" ]; then
      FIRST_AVAILABLE="$FALLBACK_SSH_KEY"
    else
      echo "No SSH keys found. Exiting."
      return 1
    fi
  fi

  if [ -z "$IS_AGENT" ]; then
    echo "SSH agent not running. Should have been bootstraped with the shell."
    echo -e "Run \`eval \"\$(ssh-agent -s)\"\` to start it. Investigate why not running."
    return 1
  fi

  IS_IDENTIFIED=$(ssh-add -l | grep "mygithubaccountemail@outlook.com")

  if [ -z "$IS_IDENTIFIED" ] && [ -f "$FIRST_AVAILABLE" ]; then
    /usr/bin/ssh-add "$FIRST_AVAILABLE" &&
      lazygit
  else
    lazygit
  fi
}

```
If anyone is better at scripting feel free to chime in with suggestions.
NOTE:  my ssh-agent is bootstraped in .zshrc to start with the shell session so the script expects it to be running. If not it prints a hint what to do.

**Comment by MykolaVaskevych** [comment]
This is off-topic, but I hate dealing with SSH keys. I use 1Password manager with a friendly CLI and SSH agent. So, all the hard work of generating, signing, and using SSH 1Password is done for me automatically. It works fine with any terminal and lazy git with zero additional setup steps. https://developer.1password.com/docs/ssh/agent

**Reply by MykolaVaskevych** [reply]
And no, I am not a bot, and it's not sponsored.

**Comment by SparkyCloudy** [comment]
Assume you already setup everything (git ssh and gpg), you can execute this command with elevated privilege (RAD, Windows builtin sudo, gsudo, etc):

```sh
git config --global core.sshCommand C:/Windows/System32/OpenSSH/ssh.exe
```
